src/problem.jl

```diff
@@ -25,101 +29,129 @@ abstract type SolverMode end
 """
    MaterializedSolve()
 
-Compute all operations on a fully materialised Z matrix.
+solve all operations on a fully materialised Z matrix.
 
 This is inneficient for CPUS, but may be best for GPUs
 using CuSSP.jl ?
 """
 struct MatrixSolve <: SolverMode end
 """
-   ColumnSolve()
+   VectorSolve()
 
-Compute all operations column by column,
+Solve all operations column by column,
 after precomputing lu decompositions.
 
 TODO we can put LinearSolve.jl solver objects in the struct
 """
-struct VectorSolve <: SolverMode end
+struct VectorSolve{T} <: SolverMode 
+    keywords::T
+end
 
 """
-    Problem(ops...; mode, θ)
+    Problem(graph_measures...; solver, θ)
 
-Combine multiple compute operations into a single object, 
+Combine multiple solve operations into a single object, 
 to be run in the same job.
 """
-@kwdef struct Problem{O,M<:SolverMode,T} <: AbstractProblem
-    ops::O
-    mode::M=VectorSolve()
-    θ::T=1.0
+@kwdef struct Problem{GM,CM<:ConnectivityMeasure,DT,SM<:SolverMode} <: AbstractProblem
+    graph_measures::GM
+    connectivity_measure::CM=LeastCostDistance()
+    distance_transformation::DT=nothing
+    solver::SM=VectorSolve((;))
 end
-Problem(ops::AbstractOperation...; kw...) = Problem(ops; kw...)
-Problem(ops::Tuple; kw...) = Problem(; ops, kw...)
-Problem(ops::NamedTuple; kw...) = Problem(; ops, kw...)
-Problem(p::AbstractProblem; mode=p.mode, θ=nothing) = Problem(o.ops, mode, θ)
+Problem(gms::GraphMeasure...; kw...) = Problem(gms; kw...)
+Problem(graph_measures::Union{Tuple,NamedTuple}; kw...) = Problem(; graph_measures, kw...)
+Problem(p::AbstractProblem; solver=p.solver, θ=nothing) = Problem(o.graph_measures, solver, θ)
+
+solve(p::Problem, rast::RasterStack) = solve(p, Grid(p, rast))
+solve(p::Problem, g::Grid) = solve(p.solver, connectivity_measure(p), p, g::Grid)
 
-compute(p::Problem, rast::RasterStack) = compute(p, Grid(p, rast))
-compute(p::Problem, g::Grid) = compute(p.mode, p, g::Grid)
 
 # Use an iterative solver so the grid is not materialised
-function compute(m::VectorSolve, p::AbstractProblem, g::Grid)
-    # Compute Z column by column
-    _, targetnodes = _targetidx_and_nodes(g)
+function solve(m::VectorSolve, cm::FundamentalMeasure, p::AbstractProblem, g::Grid)
```

Comment:
Excellent, this looks good! But I think that this should be a `DirectVectorMode`, and we likely need to implement something similar for `IndirectVectorMode`, where instead of the `init(linprob)`, we calculate a preconditioner, and then dispatch the calculations over the threads. This mode will make the solve free from any matrix materialization. It would be awesome to also have in the future an `IndirectMatrixMode`, this would work great on GPUs.

Reply:
Actually, why not call those `XxVectorMode` and `XxBatchMode`?